design/proposal-logging.md

```diff
@@ -0,0 +1,188 @@
+# Consolidate pod logs
+ 
+## Goals
+ 
+Consolidate all pod log collection to one location so that tools such as [sbctl](https://github.com/replicatedhq/sbctl) can locate them.
+ 
+## Non Goals
+ 
+As a side effect of this change, it may be that the logs collector is entirely deprecated.
+ 
+## Background
+ 
+Currently pod logs are collected by the clusterResources collector from containers for pods that have terminated with an error or are crash-looping, and stored in `/cluster-resources/pods/logs/[namespace]/[pod]/[container].log`.  The logs collector stores each collector’s pod logs in `/[name]/[pod-name]/[container-name].log`, if it is named, or at the root of the support bundle if it is not.  If a pod contains only one container, the file is named `[pod-name].log`.  This means that there are 3-4 different places where logs are to be found.
+ 
+The sbctl tool can only find pod logs if they are stored in the clusterResources directory, and so cannot return any logs collected by the logs collector.
+ 
+Some other collectors, such as Ceph and Longhorn, would benefit from pod log collection but do not appear to have any logic to do so, requiring an additional `logs` collector to be added should we wish to have those logs.
+
+Currently collectors run synchronously.
+
+## High-Level Design
+
+Introduce a new function(s) which does the work of pod log collection, storing those logs where `sbctl` can find them in `/cluster-resources/pods/logs/[namespace]/[pod]/[container].log`.
+
+Switch all collectors to use goroutines to collect (concurrently).
+
+Add a channel to each collector (including `logs`), of a type where collectors can, towards the end of their run, send a slice to the channel specifying the pods that particular collector wants to collect logs from.
+
+The new pod log collection function can read that channel, and trigger collecting logs from any pods that have not already been collected.
+
+Alter the clusterResources collector:
+
+* add an option (default none) to collect pod logs from all pods/containers according to a selector rather than just error pods
+* add an option (boolean, default false) to collect logs from all pods/containers within the namespace(s) supplied, or all namespaces if no namespace is listed
+* add an option to override the default `MaxLines` (500) currently used in the logs collector
+* Add the current `unhealthyPods` slice, plus a slice of the pods found with the new selector option, plus a slice from other collectors (e.g. Ceph, logs collectors) to a new slice, and sort/dedupe it.
```

Comment:
This all sounds great -- any chance of seeing a draft of what a yaml snippet of this would look like in a SupportBundle yaml?

Reply:
I love this suggestion having a sample input can really help drive understanding of the change.

Reply:
thanks - yeah that needs to go in here

Reply:
I've pulled back from adding this to clusterResources, simply because it is already implemented in the logs collector.  We can, and should, make sure that this stores the logs that are relevant and that might mean adding 'all' at some point.  However, the logs collector currently stores the logs in a place where sbctl can't find them, which is addressed in #744